librosa/core/pitch.py

```diff
@@ -336,3 +336,166 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
+    if fmax is None:
+        fmax = sr / 4
+
+    # Set the default hop, if it's not already specified.
+    if hop_length is None:
+        hop_length = int(frame_length // 4)
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Pad the time series so that frames are centered.
+    y = np.pad(y, int(frame_length // 2), mode=pad_mode)
+
+    # Compute autocorrelation in each frame.
+    y_frames = util.frame(y, frame_length=frame_length, hop_length=hop_length)
+    n_frames = y_frames.shape[1]
+    acf_frames = librosa.autocorrelate(y_frames, axis=0)
+
+    # Difference function.
+    min_period = np.maximum(int(np.floor(sr/fmax)), 2)
+    max_period = np.minimum(int(np.ceil(sr/fmin)), frame_length-1)
+    boxcar_window = scipy.signal.windows.boxcar(frame_length)
+    energy = np.convolve(y*y, boxcar_window, mode="same")
+    energy_frames = util.frame(
+        energy, frame_length=frame_length, hop_length=hop_length)
+    energy_0 = energy_frames[0, :]
+    energy_tau = energy_frames[(min_period-1):(max_period+1), :]
+    acf_tau = acf_frames[(min_period-1):(max_period+1), :]
+    yin_frames = energy_0 + energy_frames - 2*acf_frames
+
+    # Cumulative mean normalized difference function.
+    # NB: Equation 8 in de Cheveigné and Kawahara JASA 2002 seems to imply that
+    # the denominator is: cumsum(difference_fames)/tau_range, not
+    # cumsum(difference_frames/tau_range) as we implement it.
+    # However, going back to line 48 of the MATLAB code by AdC, we find the line:
+    # dd= d(2:end) ./ (cumsum(d(2:end)) ./ (1:(p.maxprd)));
+    # in which the parentheses indicate that the division by tau must happen
+    # before the cumulative summation, not after.
+    # Therefore, in this implementation, we purposefully diverge from Equation 8
+    # and follow the MATLAB reference implementation instead.
+    if cumulative:
+        yin_numerator = yin_frames[(min_period-1):(max_period+1), :]
+        tau_range = np.arange(1, frame_length)[:, np.newaxis]
+        cumulative_mean = np.cumsum(yin_frames[1:, :]/tau_range, axis=0)
+        epsilon = np.finfo(y.dtype).eps
+        yin_denominator = epsilon + cumulative_mean[(min_period-2):max_period, :]
+        yin_frames = yin_numerator / yin_denominator
+
+    # Parabolic interpolation.
+    # NB: perhaps these operations can be fused to alleviate memory allocation?
+    if interpolate:
+        parabola_a =\
+            (yin_frames[:-2, :]+yin_frames[2:, :]-2*yin_frames[1:-1, :]) / 2
+        parabola_b = (yin_frames[2:, :]-yin_frames[:-2, :]) / 2
+        parabolic_shifts = -parabola_b / (2*parabola_a)
+        parabolic_values =\
+            yin_frames[1:-1, :] - parabola_b*parabola_b/(4*parabola_a)
+        is_trough = np.logical_and(
+            yin_frames[1:-1, :]<yin_frames[:-2, :],
+            yin_frames[1:-1, :]<yin_frames[2:, :])
+        yin_frames = yin_frames[1:-1, :]
+        yin_frames[is_trough] = parabolic_values[is_trough]
```

Comment:
is it necessary to do this by hand, or could polyfit be used instead?

Reply:
i think polyfit would be slower.
this code vectorizes several dozens/hundreds of parabolic interpolations in the same linalg operations.
polyfit would have to be called in a loop
also note that only the troughs are interpolated. (this is in compliance with the MATLAB code)

Reply:
Ok then.  Related question: is there any redundancy between this calculation and the interpolation implemented in piptrack?  Or are they totally different?